### PR TITLE
ENH: added new output_logits option to generate function

### PR DESCRIPTION
**Pull request by mbaak**
# What does this PR do?

output_logits option behaves like output_scores, but returns the raw, unprocessed prediction logit scores, ie. the values before they undergo logit processing and/or warping. The latter happens by default for the regular output scores.

It's useful to have the unprocessed logit scores in certain circumstances. For example, unprocessed logit scores are very useful with causallm models when one wants to determine the probability of a certain answer, e.g. when asking a question with a yes/no answer. In that case getting the next-token probabilities of both "yes" and "no" (and/or their relative ratio) is of interest for classification. The reason for getting these _before_ logit processing and/or warping is b/c a) that can change the probabilities or b) reject the tokens of interest / reduce the number of tokens to just 1.

In practice this can be used to generate confidence / classification scores when eg. using causallm models for question-answering tasks. Query your language model with: "Is the {statement} correct? Answer yes or no:", take the raw logit scores and softmax them, and calculate the score: prob(yes) / (prob(yes) + prob(no)) to get a useful classification score. 

For an example use-case see paper TabLLM: Few-shot Classification of Tabular Data with Large Language Models by Stefan Hegselmann, Alejandro Buendia, Hunter Lang, Monica Agrawal, Xiaoyi Jiang, and David Sontag. https://arxiv.org/abs/2210.10723

In addition:
- added dedicated unit test: tests/generation/test_utils/test_return_unprocessed_logit_scores which tests return of logics with output_logits=True in generation.
- set output_logits=True in all other generation unit tests, that also have output_scores=True.

Fixes # (issue)
NA

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ X ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ X ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      (Yes, I've seen it discussed but now cannot refind the link.) 
- [ X ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ X ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- generate: @gante
- text models: @ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by gante** [comment]
Hi @mbaak 👋  Thank you for opening the PR 🤗 

I haven't seen a request for this feature (or I have no recollection of it). Personally, I am also not very convinced that it is very useful -- at the end of the day, the model will select the next token _after_ applying the logits processors, if there are any. Since merging this PR would increase the complexity of our codebase AND I'm not convinced, I won't be merging it for now.

However, I may have a poor perception of this feature's usefulness. I'm going to do my standard bargain: if this comment gets 20 reactions, it means there are users looking for this feature, in which case I will add the feature 🤗 (Whoever does the 20th reaction, please tag me)

**Comment by mbaak** [comment]
Hi @gante,

Let me try to explain the use-case better, my description was a bit short and perhaps unclear. 
(Indeed for next-token-generation the raw logits are not relevant.) 

I'm using a RAG setup for question-answering on large documents. (For example, a document describes a project, and a question can be: what's the location of the project?) As causallm model I'm using llama2 (13B) with default settings, meaning with logic processing and warping turned on. The answers are generated as usual, so based on the processed logits to generate the best possible next tokens.

However, these answers need to be reviewed for mistakes/hallucinations. So we want a confidence score for each answer - one that helps in the review. A causallm model does not provide that of course, just generated tokens, so we got to be creative and came up with a possible solution. 

(We find this setup with an LLM is significantly more accurate than a RAG using dedicated bert-based QA models, which do provide scores.) 
(One can prompt a causallm model to generate a confidence score, but it seems that only works for O(100)B parameter models, not for smaller ones.)

We do the following: 
After each answer the same instantiated model is queried again and asked: "Given the context is the provided answer correct? Yes or No." We are then interested in the relative probability: P(Yes|text) / (P(No|text) + P(Yes|text)). 
Interestingly, and perhaps surprisingly, in practice this ratio turns out to be a pretty reliable confidence score, ie. closer to zero is more inaccurate and closer to one is more accurate. 

But for this to work we need the unprocessed logits. B/c after warping (and using this query) normally only one token remains, yes or no, which has the (renormalized) probability 1, so the confidence score is always 0 or 1. That does not help us.

So I'm using a causallm model in a somewhat unconventional way. But one that I believe is very useful, in the sense that it provides a (much needed!) functionality that is otherwise missing for RAGs (as far as I have seen).

Hope this makes it more clear! 🙂

There may be a different way to do this, but I'm afraid I am not aware of it. Hence the PR for unprocessed logits, which for the reason above I think this would a useful addition to have.

**Comment by sbrugman** [comment]
@gante This feature was requested earlier: https://github.com/huggingface/transformers/issues/17521

**Comment by mbaak** [comment]
@sbrugman Thanks for the reminder. 
@gante Does that change you mind? :-) (In the other thread you are okay with exactly this.)

**Comment by koptagel** [comment]
This would be very beneficial for my current project, I would love this PR to be merged.

**Comment by hlsnoek** [comment]
upvote

**Comment by gante** [comment]
It seems I was wrong, and several people do want it :) Reverting my decision, I'll review the PR so as to include the feature

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -306,6 +309,7 @@ def __init__(self, **kwargs):
         self.output_attentions = kwargs.pop("output_attentions", False)
         self.output_hidden_states = kwargs.pop("output_hidden_states", False)
         self.output_scores = kwargs.pop("output_scores", False)
+        self.output_logits = kwargs.pop("output_logits", False)
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:312]:
```suggestion
        self.output_logits = kwargs.pop("output_logits", None)
```
We'd rather default to unset (`None`) than to a valid option (`False`) on all new attributes. That way, if we want to change the default in the future, we can easily know when a user has intentionally set this attribute :)

File: src/transformers/generation/configuration_utils.py [hunk 2]
```diff
@@ -216,6 +216,9 @@ class GenerationConfig(PushToHubMixin):
             more details.
         output_scores (`bool`, *optional*, defaults to `False`):
             Whether or not to return the prediction scores. See `scores` under returned tensors for more details.
+        output_logits (`bool`, *optional*, defaults to `False`):
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:219]:
```suggestion
        output_logits (`bool`, *optional*):
```
(see related comment below)

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1769,6 +1798,9 @@ def contrastive_search(
                 for more details.
             output_scores (`bool`, *optional*, defaults to `False`):
                 Whether or not to return the prediction scores. See `scores` under returned tensors for more details.
+            output_logits (`bool`, *optional*, defaults to `False`):
+                Whether or not to return the raw prediction logit scores. See `raw_logits` under returned tensors
```
Comment (gante) [inline src/transformers/generation/utils.py:1802]:
```suggestion
                Whether or not to return the raw prediction logit scores. See `logits` under returned tensors
```
`logits` is the tensor name defined in the output classes. This pattern occurs more times in this diff, make sure to correct them :)

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -2318,15 +2362,17 @@ def greedy_search(
             if synced_gpus and this_peer_finished:
                 continue  # don't waste resources running the code we don't need
 
-            next_token_logits = outputs.logits[:, -1, :]
+            next_tokens_logits = outputs.logits[:, -1, :]
```
Comment (gante) [inline src/transformers/generation/utils.py:2365]:
```suggestion
            next_token_logits = outputs.logits[:, -1, :]
```
logits (plural) regarding the next token (singular)

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -2318,15 +2362,17 @@ def greedy_search(
...
 
             # pre-process distribution
-            next_tokens_scores = logits_processor(input_ids, next_token_logits)
+            next_tokens_scores = logits_processor(input_ids, next_tokens_logits)
```
Comment (gante) [inline src/transformers/generation/utils.py:2368]:
```suggestion
            next_tokens_scores = logits_processor(input_ids, next_token_logits)
```

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -2318,15 +2362,17 @@ def greedy_search(
...
 
             # Store scores, attentions and hidden_states when required
             if return_dict_in_generate:
                 if output_scores:
                     scores += (next_tokens_scores,)
+                if output_logits:
+                    raw_logits += (next_tokens_logits,)
```
Comment (gante) [inline src/transformers/generation/utils.py:2375]:
```suggestion
                    raw_logits += (next_token_logits,)
```

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -3454,3 +3465,41 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
             encoder_outputs=encoder_outputs,
         )
         self.assertListEqual(outputs_assisted.tolist(), outputs_foo.tolist())
+
+    def test_return_unprocessed_logit_scores(self):
```
Comment (gante) [inline tests/generation/test_utils.py:3469]: I'd redo this test, to verify `output_logits`'s main property: that the logits match the forward pass. e.g. we could generate with `max_new_tokens=1, do_sample=True` and confirm that the returned logits match the forward pass :)

**Review by gante — COMMENTED** [review]
Thank you for the contribution 💛 

I've commented a few suggestions, they should be simple to implement!

**Comment by mbaak** [comment]
Thanks for the comments! I'll pick it up.

File: tests/generation/test_utils.py (hunk 7 above)
Comment (mbaak) [reply tests/generation/test_utils.py:3503]: @gante Thanks again for your feedback, I've implemented all of it.
I've implemented an extra unit test doing as you suggested, called test_compare_unprocessed_logit_scores().
I've also kept the original test in place, as it illustrates how one can use the output_logits option in practice.

**Comment by mbaak** [comment]
@gante I've implemented your feedback, the code should be good to go I think!

**Review by gante — APPROVED** [review]
Thank you for iterating 🤗

**Comment by gante** [comment]
@mbaak To make our CI go green, you will need to:
1. rebase with the latest `main`
2. run `make fixup` on your terminal, within the transformers folder
3. force push the changes

**Comment by mbaak** [comment]
@gante Done!

**Comment by gante** [comment]
@mbaak there seems a missing `output_logits` somewhere in the code, CI is complaining :D

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this feature - LGTM! 

Just some small nit comments to address, but otherwise looks ready for merge

File: tests/generation/test_utils.py [hunk 8]
```diff
@@ -3476,3 +3487,71 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
...
+        # Unprocessed logit scores are very useful with causallm models when one wants to determine the probability of
+        # a certain answer, for example when asking a question with a yes/no answer. In that case getting the next-token
+        # probabilities of both "yes" and "no" (and/or their relative ratio) is of interest for classification.
+        # The reason for getting these _before_ logit processing is b/c a) that can warp the probabilities or b)
+        # reject the tokens of interest / reduce the number of tokens to just 1.
+        # For an example use-case see paper TabLLM: Few-shot Classification of Tabular Data with Large Language Models
+        # by Stefan Hegselmann, Alejandro Buendia, Hunter Lang, Monica Agrawal, Xiaoyi Jiang, and David Sontag.
+        # https://arxiv.org/abs/2210.10723
+
```
[the remark this answers is not in the document]
Comment (mbaak) [reply tests/generation/test_utils.py:3530]: Hi, if you have a suggestion where to put an example of this feature let me know. For example, how to use a CausalLM model to get a classification score to a yes/no question - I'm happy to add that somewhere.

[the remark this answers is not in the document]
Comment (mbaak) [reply tests/generation/test_utils.py:3530]: Else I will write a blog about it on medium :-)

File: tests/generation/test_utils.py [hunk 9]
```diff
@@ -3476,3 +3487,71 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
             encoder_outputs=encoder_outputs,
         )
         self.assertListEqual(outputs_assisted.tolist(), outputs_foo.tolist())
+
+    def test_compare_unprocessed_logit_scores(self):
+        # Get unprocessed logit scores back from model generate function.
+        # Assert that unprocessed logits from generate() are same as those from modal eval()
+        # For an example application of those logit score see example below.
+        # test_access_unprocessed_logit_scores()
+
```
[the remark this answers is not in the document]
Comment (mbaak) [reply tests/generation/test_utils.py]: removed

File: tests/generation/test_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mbaak) [reply tests/generation/test_utils.py]: removed

**Comment by mbaak** [comment]
@gante Fyi I'm working on fixing the tests. (By default they're skipped locally, that's why I missed them earlier.)

**Comment by mbaak** [comment]
@gante I fixed the CI tests (I had only run the generation/ tests locally, not the model ones), and have implemented @amyeroberts' feedback. I think it's good to go now!

**Comment by gante** [comment]
@mbaak thank you for iterating and making the library richer 💛
